Enable Full Screen Splash Images

bebon901 opened the pull request:
This will allow users to enter a cmdline value which will enable the kernel to read a PPM file, and then rendering it fullscreen at the early boot stage.
I am aware it needs a _lot_ of tidying up, but put here for others to reference!

pelwell commented on drivers/video/fbdev/core/fb_logo.c:
```diff
@@ -275,84 +319,544 @@ static void fb_do_show_logo(struct fb_info *info, struct fb_image *image,
+
+__setup("fullscreen_logo=", fullscreen_logo_setup);
+
+ssize_t read_file_from_initramfs(const char *filepath, char **buffer, size_t max_size)
```
Why make the name initramfs-specific? Doesn't this work just as well with files in a regular FS?

pelwell commented on drivers/video/fbdev/core/fb_logo.c:
```diff
@@ -275,84 +319,544 @@ static void fb_do_show_logo(struct fb_info *info, struct fb_image *image,
+    ret = bytes_read; // Return actual bytes read
+
+out_free:
+    if (ret < 0 && *buffer) {
```
We already know that `*buffer` is non-NULL

pelwell commented on drivers/video/fbdev/core/fb_logo.c:
```diff
@@ -275,84 +319,544 @@ static void fb_do_show_logo(struct fb_info *info, struct fb_image *image,
+
+    // 2. Allocate buffer
+	printk(KERN_ALERT "File Length: %d", file_len);
+    *buffer = kvmalloc(file_len + 1, GFP_KERNEL); // +1 for null terminator
```
It would be nicer to use a temporary pointer for the allocated buffer, and only assign to `*buffer` in the success path.

pelwell commented on drivers/video/fbdev/core/fb_logo.c:
```diff
@@ -275,84 +319,544 @@ static void fb_do_show_logo(struct fb_info *info, struct fb_image *image,
+		printk(KERN_INFO "Attempting to read file from initramfs... --------------------------------------------------------------------------------------------------\n");
+
+		long width = 0, height = 0, max_color_value = 0;
+		long image_palette[224][3];
```
It would be nice to declare a typedef for this type and use it throughout.

pelwell reviewed (commented):
I'm not going to give this a full review until you've tidied it up some - it's too tiring otherwise.

A few kernel coding rules (see https://www.kernel.org/doc/html/v6.12/process/coding-style.html for more):
1. Use TABs for indentation, with tab stops set to 8. Spaces can be used to precise alignment, but only after any TABs.
2. You have the preferred brace style and no trailing whitespace, so that's good.
3. Your local variable declarations are also pretty good, but for full marks put them in "reverse Christmas tree" format, longest to shortest, making sure there is a blank line before any statements.
4. Make all local-scope functions `static`, and ideally with a module-specific prefix.
5. Prefer `pr_info` (or `dev_info`) over `printk(KERN_INFO `.

`scripts/checkpatch.pl -g <from-commit>..<to-commit>` is your fussy friend.

Commit:
Fullscreen Splash Images

Enable by adding the following to cmdline.txt:
`fullscreen_logo_name=logo.tga fullscreen_logo=1`
Will show the logo file present in /lib/firmware/ on the screen.
This will be fullscreen and rendered early at boot.
Any remaining space is filled with solid color from the image border.
If TGA file is too big, image is clipped accordingly.

Signed-off-by: Ben Benson <ben.benson@raspberrypi.com>

pelwell commented:
This is looking much better. You could replace the typedef with a structure whose only member is the two-dimensional array. And on that subject, is there any reason not to make it an array of `u8`/`uint8_t`s? Using longs is wasteful and confusing, in the sense that it makes the reader wonder where you use values large enough to justify their use.

pelwell commented on drivers/video/fbdev/core/fb_logo.c:
```diff
@@ -275,6 +304,164 @@ static void fb_do_show_logo(struct fb_info *info, struct fb_image *image,
+{
+	fullscreen_logo_enabled = 1;
+	fullscreen_logo_path = str;
+	pr_info("Fullscreen Logo Enabled: %d", fullscreen_logo_enabled);
```
This will only ever show "Fullscreen Logo Enabled: 1", so I think it could be combined with the "Path" line.

pelwell commented on drivers/video/fbdev/core/fb_logo.c:
```diff
@@ -275,6 +304,164 @@ static void fb_do_show_logo(struct fb_info *info, struct fb_image *image,
+		const char *current_ptr = file_content;
+		const char *end_ptr = file_content + len;
+
+		if (len < 18) {
```
I don't understand why checkpatch hasn't complained about this code before the second group of declarations. You could move it down to before the image_type check, or merge it with the other `len` error case after this code block.

pelwell commented on drivers/video/fbdev/core/fb_logo.c:
```diff
@@ -275,6 +304,164 @@ static void fb_do_show_logo(struct fb_info *info, struct fb_image *image,
+		const char *end_ptr = file_content + len;
+
+		if (len < 18) {
+			pr_info("Invalid logo file: TGA file too small for header\n");
```
Error messages should be output at `ERROR` (`pr_err`) or `WARN` (`pr_warn`) level to make it easier to find them (e.g. `dmesg -l err,warn`). This applies in several places in this file.

pelwell commented on drivers/video/fbdev/core/fb_logo.c:
```diff
@@ -275,6 +304,164 @@ static void fb_do_show_logo(struct fb_info *info, struct fb_image *image,
+		current_ptr = file_content + 18 + id_length;
+
+		read_logo = kmalloc_array(width, height, GFP_KERNEL);
+		if (read_logo == NULL)
```
`if (!read_logo)` is the preferred form in the kernel.

pelwell commented on drivers/video/fbdev/core/fb_logo.c:
```diff
@@ -275,6 +304,164 @@ static void fb_do_show_logo(struct fb_info *info, struct fb_image *image,
+
+		// TGA pixels are stored bottom-to-top by default, unless bit 5 of
+		// image_descriptor is set
+		bool top_to_bottom = (image_descriptor & 0x20) != 0;
```
More mid-block declarations...

pelwell commented on drivers/video/fbdev/core/fb_logo.c:
```diff
@@ -275,6 +304,164 @@ static void fb_do_show_logo(struct fb_info *info, struct fb_image *image,
+		// Skip header + ID field
+		current_ptr = file_content + 18 + id_length;
+
+		read_logo = kmalloc_array(width, height, GFP_KERNEL);
```
This will need to be freed in the error paths. It's common to end a function with a list of labelled cleanup steps in reverse order, with error handling code jumping to the appropriate one.

pelwell reviewed (commented):
Just a few last-minute suggestions. I nearly hit the Merge button, but this is so close to being good...
